include/pybind11/detail/init.h

```diff
@@ -0,0 +1,333 @@
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "../class_support.h"
+
+NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+NAMESPACE_BEGIN(detail)
+NAMESPACE_BEGIN(initimpl)
+
+inline void no_nullptr(void *ptr) {
+    if (!ptr) throw type_error("pybind11::init(): factory function returned nullptr");
+}
+
+// Makes sure the `value` for the given value_and_holder is not preallocated (e.g. by a previous
+// old-style placement new `__init__` that requires a preallocated, uninitialized value).  If
+// preallocated, deallocate.  Returns the (null) value pointer reference ready for allocation.
+inline void *&deallocate(value_and_holder &v_h) {
+    if (v_h) v_h.type->dealloc(v_h);
+    return v_h.value_ptr();
+}
+
+PYBIND11_NOINLINE inline value_and_holder load_v_h(handle self_, type_info *tinfo) {
+    if (!self_ || !tinfo)
+        throw type_error("__init__(self, ...) called with invalid `self` argument");
+
+    auto *inst = reinterpret_cast<instance *>(self_.ptr());
+    auto result = inst->get_value_and_holder(tinfo, false);
+    if (!result.inst)
+        throw type_error("__init__(self, ...) called with invalid `self` argument");
+
+    return result;
+}
+
+
+// Implementing functions for all forms of py::init<...> and py::init(...)
+template <typename Class> using Cpp = typename Class::type;
+template <typename Class> using Alias = typename Class::type_alias;
+template <typename Class> using Holder = typename Class::holder_type;
+
+// Takes a Cpp pointer and returns true if it actually is a polymorphic Alias instance.
+template <typename Class, enable_if_t<Class::has_alias, int> = 0>
+static bool is_alias(Cpp<Class> *ptr) {
+    return dynamic_cast<Alias<Class> *>(ptr) != nullptr;
+}
+// Failing fallback version of the above for a no-alias class (always returns false)
+template <typename Class>
+constexpr static bool is_alias(void *) { return false; }
+
+// Attempts to constructs an alias using a `Alias(Cpp &&)` constructor.  This allows types with
+// an alias to provide only a single Cpp factory function as long as the Alias can be
+// constructed from an rvalue reference of the base Cpp type.  This means that Alias classes
+// can, when appropriate, simply define a `Alias(Cpp &&)` constructor rather than needing to
+// inherit all the base class constructors.
+template <typename Class, enable_if_t<std::is_constructible<Alias<Class>, Cpp<Class> &&>::value, int> = 0>
+static bool construct_alias_from_cpp(value_and_holder &v_h, Cpp<Class> &&base) {
+    deallocate(v_h) = new Alias<Class>(std::move(base));
+    return true;
+}
+// Fallback of the above for aliases without an `Alias(Cpp &&) constructor; does nothing and
+// returns false to signal the non-availability of the special constructor.
+template <typename Class, enable_if_t<!std::is_constructible<Alias<Class>, Cpp<Class> &&>::value, int> = 0>
+static constexpr bool construct_alias_from_cpp(value_and_holder &, Cpp<Class> &&) { return false; }
```

Comment:
The control flow built around this feels a bit more complicated than it should be. This `return false` results in two different `throw type_error` messages below, but they have the same cause. Consider switching to `void` functions and place the `throw` statement in the `false` fallback. (I'm also using simple tag dispatch in the example below because it's a nice lightweight option, especially for binary overloads, and it also nicely maps to the future `if constexpr`.)

```c++
template <typename Class>
using is_alias_constructible = std::is_constructible<Alias<Class>, Cpp<Class> &&>;

template <typename Class>
void construct_alias_from_cpp(std::true_type /*is_alias_constructible*/,
                              value_and_holder &v_h, Cpp<Class> &&base) {
    deallocate(v_h) = new Alias<Class>(std::move(base));
}

template <typename Class>
[[noreturn]] void construct_alias_from_cpp(std::false_type /*is_alias_constructible*/,
                                           value_and_holder &, Cpp<Class> &&) {
    throw type_error("pybind11::init(): unable to convert returned instance to required "
                     "alias class: no `Alias<Class>(Class &&)` constructor available");
}
```

```c++
if (Class::has_alias && need_alias && !is_alias<Class>(ptr)) {
    ...
    construct_alias_from_cpp<Class>(is_alias_constructible<Class>{}, v_h, std::move(*ptr));
} else {
    deallocate(v_h) = ptr;
}
```